### ISSUE TITLE
Why is the quantified TF_LITE slower than the unquantified TF_LITE test time on the PC side？

### ISSUE THREAD

**Issue by huafeihuayu** [tensorflow/tensorflow#43013]
import tensorflow as tf
import numpy as np
import cv2
import os
import time
from PIL import Image

model = "D:/sfz/tf_sfz/uint9.tflite"
interpreter = tf.lite.Interpreter(model_path=model)
interpreter.allocate_tensors()
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()
print(str(input_details), '\n', str(output_details))
img = Image.open(“D:/data/000001.jpg”)
        num = img.size[0] * 32 // img.size[1]
        if num > max_width:
            num = max_width
        if num < 5:
            num = 5
        out_mid = img.resize((num, 32))
        out = img.resize((max_width, 32))
        out_crop = out_mid.crop((num - 1, 0, out_mid.size[0], out_mid.size[1]))
        out.paste(out_mid, (0, 0))
        while num < max_width:
            out.paste(out_crop, (num, 0))
            num += 1
        img = cv2.cvtColor(np.asarray(out), cv2.COLOR_RGB2BGR)
        inputImg=(img/255).astype(np.float32)
        interpreter = tf.lite.Interpreter(model_path=model)
        interpreter.allocate_tensors()
        input_details = interpreter.get_input_details()
        output_details = interpreter.get_output_details()
        interpreter.set_tensor(input_details[0]['index'], [inputImg])
        a = time.time()
        # print(a)
        interpreter.invoke()
        b = time.time()
        # print(b)
        print((b - a) * 1000)

**Comment by ravikyram** [comment]
@huafeihuayu 

Can you please fill issue template.
Request you to share colab link or simple standalone code to reproduce the issue in our environment.It helps us in localization issue faster.Thanks!

**Comment by huafeihuayu** [comment]
> @huafeihuayu
> 
> Can you please fill issue template.
> Request you to share colab link or simple standalone code to reproduce the issue in our environment.It helps us in localization issue faster.Thanks!

@ravikyram 
I've provided the code

**Comment by thaink** [comment]
@huafeihuayu 
Since you are using python, I am supposed you are running on PC.
The quantized model don't get much performance benefit on PC since it does not have specialized instruction for that.

Could you try to  compare the result on mobile. You can profile your model using one of following approaches:
https://www.tensorflow.org/lite/performance/best_practices#profile_your_model

**Comment by huafeihuayu** [comment]
> @huafeihuayu
> 因为您使用的是python，所以我认为您正在PC上运行。
> 量化模型在PC上没有太大的性能优势，因为它没有专门的说明。
> 
> 您可以尝试在移动设备上比较结果吗？您可以使用以下方法之一来分析模型：[https](https://www.tensorflow.org/lite/performance/best_practices#profile_your_model) :
> [//www.tensorflow.org/lite/performance/best_practices#profile_your_model](https://www.tensorflow.org/lite/performance/best_practices#profile_your_model)


> @huafeihuayu
> Since you are using python, I am supposed you are running on PC.
> The quantized model don't get much performance benefit on PC since it does not have specialized instruction for that.
> 
> Could you try to compare the result on mobile. You can profile your model using one of following approaches:
> https://www.tensorflow.org/lite/performance/best_practices#profile_your_model

@thaink @ravikyram 
import tensorflow as tf
saved_model_dir='D:/tf_ocr/tf_resnet7'
converter = tf.lite.TFLiteConverter.from_saved_model(saved_model_dir)
converter.supported_types = [tf.lite.constants.INT8]
converter.optimizations = [tf.lite.Optimize.DEFAULT]
Tflite_quanit_model = converter.convert()
open("D:/sfz/tf_sfz/model10.tflite","wb").write(Tflite_quanit_model)

Since it is trainable quantized, using savermodel.pb to quantify the TF_lite model input/output or Float32 instead of UINT8, does this decrease the reasoning speed on the move?

**Comment by thaink** [comment]
What do you mean by "trainable quantized"?
If you use quantization aware training, the conversion should be as simple as:
https://www.tensorflow.org/model_optimization/guide/quantization/training_example#create_quantized_model_for_tflite_backend

**Comment by huafeihuayu** [comment]
> What do you mean by "trainable quantized"?
> If you use quantization aware training, the conversion should be as simple as:
> https://www.tensorflow.org/model_optimization/guide/quantization/training_example#create_quantized_model_for_tflite_backend

@thaink @ravikyram 
import tensorflow as tf
saved_model_dir='D:/sfz/tf_sfz'
def representative_data_gen():
for input_value in tf.data.Dataset.from_tensor_slices(train_images).batch(1).take(100):
yield [input_value]

converter = tf.lite.TFLiteConverter.from_saved_model(saved_model_dir)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.representative_dataset = representative_data_gen
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.inference_input_type = tf.uint8
converter.inference_output_type = tf.uint8
tflite_model_quant = converter.convert()
open("D:/sfz/tf_sfz/model25.tflite","wb").write(tflite_model_quant)

the train_images don't know what it is,I only know to select a part of the representative data, I don't know what to do next?The training data is JPG file and TXT file，They are in a folder, how do I make train_images?What type of train_images is it?Can you fix the code a little bit

**Comment by thaink** [comment]
You don't need to have accuracy data for representative_dataset so just images are enough.
Assumes image_paths is the list of your image files. You can take a look at following example:
def representative_dataset_gen(image_paths):
	image_paths_tensor = tf.convert_to_tensor(image_paths, dtype=tf.string)
	dataset = tf.data.Dataset.from_tensor_slices(image_paths_tensor)

	def map_fn(path):
	    image = tf.image.decode_jpeg(tf.read_file(path))
	    image = tf.image.resize_images(out_shape)
	    image = tf.to_float(image) /127.5 - 127.5
	    return image

	dataset = dataset.map(map_fn)
	dataset = dataset.batch(batch_size).take(100)

	iterator = dataset.make_one_shot_iterator()
	for _ in range(number_of_calibration_steps):
	    images = iterator.get_next()
	    yield [images]

Please make sure the normalization step in map_fn is correct.

**Comment by ymodak** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
